infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/utils/UpgradeManager.java

```diff
@@ -0,0 +1,213 @@
+package org.corfudb.infrastructure.logreplication.utils;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.infrastructure.LRRollingUpgradeHandler;
+import org.corfudb.infrastructure.logreplication.infrastructure.plugins.ILogReplicationVersionAdapter;
+import org.corfudb.infrastructure.logreplication.infrastructure.plugins.LogReplicationPluginConfig;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.collections.CorfuStore;
+import org.corfudb.runtime.collections.CorfuStoreEntry;
+import org.corfudb.runtime.collections.Table;
+import org.corfudb.runtime.collections.TableOptions;
+import org.corfudb.runtime.collections.TxnContext;
+import org.corfudb.runtime.exceptions.TransactionAbortedException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.IntervalRetry;
+import org.corfudb.util.retry.RetryNeededException;
+import org.corfudb.utils.CommonTypes.Uuid;
+import org.corfudb.utils.LogReplicationStreams.Version;
+import org.corfudb.utils.LogReplicationStreams.VersionString;
+
+import java.io.File;
+import java.lang.reflect.InvocationTargetException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.NoSuchElementException;
+
+import static org.corfudb.runtime.view.TableRegistry.CORFU_SYSTEM_NAMESPACE;
+
+/**
+ * Manages version and upgrade-related metadata and provides utility methods for the same.
+ */
+@Slf4j
+public class UpgradeManager {
+    public static final String LOG_REPLICATION_PLUGIN_VERSION_TABLE = "LogReplicationPluginVersion";
+
+    // TODO: During a rolling upgrade, all nodes will not be on the same version.  This key type must change to
+    //  reflect the version per node.
+    public static final String VERSION_PLUGIN_KEY = "VERSION";
+
+    private final VersionString versionString = VersionString.newBuilder().setName(VERSION_PLUGIN_KEY).build();
+
+    private final CorfuStore corfuStore;
+
+    @Getter
+    private LRRollingUpgradeHandler lrRollingUpgradeHandler;
+
+    private ILogReplicationVersionAdapter logReplicationVersionAdapter;
+
+    private final String pluginConfigFilePath;
+
+    @Getter
+    private static String currentVersion;
```

Comment:
can we add a comment to say this value is the _node_ version extracted from the plugin?

Reply:
added the comment and renamed the variable to nodeVersion